File: MyScript/edk/java/examples/writeradaptation/ConjointRecognitionAndAdaptation.java
package writeradaptation;

import java.io.File;
import java.io.IOException;

import certificates.MyCertificate;

import com.myscript.engine.Engine;
import com.myscript.engine.EngineObject;
import com.myscript.engine.FloatStructuredInput;
import com.myscript.engine.IFloatStroke;
import com.myscript.text.Resource;
import com.myscript.text.StructuredInputRecognizer;
import com.myscript.writeradaptation.WriterAdaptationCollector;
import com.myscript.writeradaptation.WriterAdaptationProfile;
import com.myscript.writeradaptation.WriterAdaptationTrainer;
import common.CFG;
import common.Ink;
import common.TextUIProgressBar;

/**
 * The <code>ConjointRecognitionAndAdaptation</code> example code shows you how
 * to perform conjoint handwriting recognition and adaptation.
 */
public class ConjointRecognitionAndAdaptation
{
  /**
   * The file where the writer adaptation profile will be stored.
   */
  private static final String PROFILE_FILE = CFG.OUTPUT_DIR + "/ConjointRecognitionAndAdaptation-profile.res";
  
  /**
   * Main entry point for the <code>ConjointRecognitionAndAdaptation</code> example program.
   *
   * @param args command line arguments, ignored.
   * @throws IOException 
   */
  public static final void main(String[] args) throws IOException
  {
    if (PROFILE_FILE == null || PROFILE_FILE.length() == 0)
    {
      System.err.println("/!\\ please choose a valid filename for writer adaptation profile file. You can choose profile filename by editing PROFILE_FILE static member variable of ConjointRecognitionAndAdaptation class");
      System.exit(-1);
    }

    System.out.println("1) creating a new Engine initialized with your certificate");
    final Engine engine = Engine.create(MyCertificate.getBytes());
    System.out.println(" . Engine created successfully");
    System.out.println();

    System.out.println("2) setting up a WriterAdaptationProfile");
    WriterAdaptationProfile profile = null;
    if (new File(PROFILE_FILE).exists())
    {
      System.out.println(" . found existing profile, loading it");
      profile = (WriterAdaptationProfile)EngineObject.load(engine, PROFILE_FILE);
    }
    else
    {
      System.out.println(" . no existing profile found, creating it");
      profile = WriterAdaptationProfile.create(engine);
    }
    System.out.println(" . profile set up successfully");
    System.out.println();

    System.out.println("3) creating and configuring a new StructuredInputRecognizer");
    final StructuredInputRecognizer recognizer = StructuredInputRecognizer.create(engine);
    System.out.println(" . recognizer created successfully");
    final Resource ak = (Resource)EngineObject.load(engine, CFG.getAkRes());
    System.out.println(" . AlphabetKnowledge resource loaded successfully");
    recognizer.attach(ak);
    System.out.println(" . AlphabetKnowledge resource attached successfully to the recognizer");
    final Resource lk = (Resource)EngineObject.load(engine, CFG.getLkTextRes());
    System.out.println(" . Main LinguisticKnowledge resource loaded successfully");
    recognizer.attach(lk);
    System.out.println(" . Main LinguisticKnowledge resource attached successfully to the recognizer");
    System.out.println(" . disposing of the main LinguisticKnowledge resource (not needed anymore)");
    lk.dispose();
    System.out.println(" . Main LinguisticKnowledge resource disposed of successfully");
    if (!CFG.getLang().contains("en"))
    {
      final String path;

      if (CFG.getLang().contains("zh") || CFG.getLang().contains("ja") || CFG.getLang().contains("ko"))
        path = CFG.getLkEnglishSecondary2Res();
      else if (CFG.getLang().contains("hi"))
        path = CFG.getLkEnglishSecondary3Res();
      else
        path = CFG.getLkEnglishSecondary1Res();

      final Resource lkEnglishSecondary = (Resource)EngineObject.load(engine, path);
      System.out.println(" . Secondary English LinguisticKnowledge resource loaded successfully");
      recognizer.attach(lkEnglishSecondary);
      System.out.println(" . Secondary English LinguisticKnowledge resource attached successfully");
      System.out.println(" . disposing of the secondary LinguisticKnowledge resource (not needed anymore)");
      lkEnglishSecondary.dispose();
      System.out.println(" . Secondary LinguisticKnowledge resource disposed of successfully");
    }
    System.out.println(" . recognizer configured successfully");
    System.out.println();

    System.out.println("4) creating a new WriterAdaptationCollector");
    final WriterAdaptationCollector collector = WriterAdaptationCollector.create(engine);
    System.out.println(" . collector created successfully");
    System.out.println();

    System.out.println("5) attaching the profile to the collector");
    collector.attach(profile);
    System.out.println(" . profile attached successfully to the collector");
    System.out.println("   . the collector will store writer data into the profile for further training");
    System.out.println();

    System.out.println("6) attaching the collector to the recognizer");
    recognizer.attach(collector);
    System.out.println(" . collector attached successfully to the recognizer");
    System.out.println("   . the recognizer will transfer incoming writer data to the collector");
    System.out.println();

    System.out.println("7) attaching the profile to the recognizer");
    recognizer.attach(profile);
    System.out.println(" . profile attached successfully to the recognizer");
    System.out.println("   . the recognizer will use existing writer data to adapt the recognition process");
    System.out.println();

    System.out.println("8) sending digital ink to the recognizer");
    final FloatStructuredInput input = FloatStructuredInput.create(engine);
    final File documentInkFile = new File(CFG.DATA_DIR + "/sentence.ink");
    Ink.parse(documentInkFile, new Ink.IParseHandler()
    {
      public void addStroke(IFloatStroke stroke)
      {
        input.addStroke(stroke);
      }
    });
    recognizer.setSource(input);
    System.out.println(" . digital ink successfully sent to the recognizer");
    System.out.println();

    System.out.println("9) running the conjoint handwriting recognition/adaptation process");
    TextUIProgressBar progressBar = new TextUIProgressBar(System.out);
    progressBar.update(0, 100);   // prints 0% status
    recognizer.run(progressBar);
    progressBar.update(100, 100); // prints 100% status
    System.out.println();
    System.out.println(" . handwriting recognition/adaptation performed successfully");
    System.out.println();

    System.out.println("10) committing the collected data");
    collector.commit();
    System.out.println(" . collected data committed successfully");
    System.out.println();

    System.out.println("11) creating and configuring a new WriterAdaptationTrainer");
    final WriterAdaptationTrainer trainer = WriterAdaptationTrainer.create(engine);
    System.out.println(" . trainer created successfully");
    trainer.attach(ak);
    System.out.println(" . AlphabetKnowledge resource attached successfully to the trainer");
    System.out.println(" . disposing of the AlphabetKnowledge resource (not needed anymore)");
    ak.dispose();
    System.out.println(" . AlphabetKnowledge resource disposed of successfully");
    trainer.attach(profile);
    System.out.println(" . profile attached successfully to the trainer");
    System.out.println(" . trainer created and configured successfully");
    System.out.println();

    System.out.println("12) running the training process on the collected data");
    TextUIProgressBar progressBarTrain = new TextUIProgressBar(System.out);
    progressBarTrain.update(0, 100);   // prints 0% status
    trainer.run(progressBarTrain);
    progressBarTrain.update(100, 100); // prints 100% status
    System.out.println();
    System.out.println(" . the trainer updated the collected writer data; next recognition will adjust itself to the writer writing style");
    System.out.println();

    System.out.println("13) storing the profile to disk");
    profile.store(PROFILE_FILE);
    System.out.println(" . profile " + new File(PROFILE_FILE).getCanonicalPath() + " stored successfully");
    System.out.println();

    // do not forget the 1st rule: you create it, you dispose of it!
    System.out.println("12) cleaning-up: disposing of objects and disposing of the Engine");
    input.dispose();
    trainer.dispose();
    collector.dispose();
    profile.dispose();
    recognizer.dispose();
    engine.dispose();
    System.out.println(" . clean-up done successfully");
    System.out.println();

    System.out.println("end of example:");
    System.out.println(" you just learnt how to perform writer adaptation");
  }

} // ConjointRecognitionAndAdaptation

File: MyScript/edk/c/examples/writeradaptation/ConjointRecognitionAndAdaptation.c
/**
 * The `ConjointRecognitionAndAdaptation` example shows you how how to perform
 * conjoint handwriting recognition and adaptation.
 */


// -- configuration ------------------------------------------------------------

// common configuration parameters
#include "../common/config.h"

// ink parsing
#include "../common/parseInk.h"

// include the MyScript Engine API header(s)
#include <c/MyScriptEngine.h>
#include <c/MyScriptText.h>
#include <c/MyScriptWriterAdaptation.h>

// include the certificate
#include "../certificates/MyCertificate.h"


// -- example body -------------------------------------------------------------

#include <stdio.h>

// global engine (for the sake of readability)
voEngine engine = NULL;

// common utility functions
#include "../common/common.h"

// path to the file that represents the writer adaptation profile
#define PROFILE_FILE OUTPUT_DIR "/ConjointRecognitionAndAdaptation-profile.res"


// -- parseInk() callback ------------------------------------------------------

struct AppContext
{
  voEngine engine;
  voFloatStructuredInput input;
};

static int parseInkCallback(const float* x, size_t xByteStride,
                            const float* y, size_t yByteStride,
                            int pointCount,
                            void* userParam)
{
  bool successful;
  struct AppContext* context = (struct AppContext*)userParam;

  successful = VO_IFACE(engine, IFloatStructuredInput)->addStroke(context->engine, context->input, x, xByteStride, y, yByteStride, pointCount);

  return successful ? 0 : voGetError(context->engine);
}


// -- main function ------------------------------------------------------------

int main()
{
  voWriterAdaptationProfile profile = NULL;
  voWriterAdaptationCollector collector = NULL;
  voWriterAdaptationTrainer trainer = NULL;
  voStructuredInputRecognizer recognizer = NULL;
  voInput input = NULL;
  voResource ak = NULL;
  voResource lk = NULL;
  voResource lkEnglishSecondary = NULL;
  TextUIProgressBar progressBar = { 0 };

  // -- check required settings have been edited -------------------------------
  if (strlen(LANG) == 0)
  {
    fprintf(stderr, "/!\\ please edit the common/config.h file and set:\n");
    fprintf(stderr, " . the LANG variable\n");
    exit(EXIT_FAILURE);
  }

  if (strlen(PROFILE_FILE) == 0)
  {
    printf("/!\\ please edit this example program setting the PROFILE_FILE variable to a valid file name\n");
    exit(EXIT_FAILURE);
  }

  // -- initialization ---------------------------------------------------------
  if (!loadMyScriptEngine(MYSCRIPT_ENGINE_LIBRARY_DIR, NULL))
    return (fprintf(stderr, "failed to load the MyScriptEngine library\n"), EXIT_FAILURE);

  // -- example code -----------------------------------------------------------
  printf("1) creating a new Engine initialized with your certificate\n");
  engine = voCreateEngine(VO_MSE_VER, &myCertificate, NULL);
  verify(engine != NULL);
  printf(" . Engine created successfully\n");
  printf("\n");

  printf("2) setting up a WriterAdaptationProfile\n");
  if (fileExists(PROFILE_FILE))
  {
    printf(" . found existing profile, loading it\n");
    profile = loadEngineObject(PROFILE_FILE);
  }
  else
  {
    printf(" . no existing profile found, creating it\n");
    profile = voCreateObject(engine, VO_WriterAdaptationProfile);
  }
  verify(profile != NULL);
  printf(" . profile set up successfully\n");
  printf("\n");

  printf("3) creating and configuring a new StructuredInputRecognizer\n");
  recognizer = voCreateStructuredInputRecognizer(engine);
  verify(recognizer != NULL);
  printf(" . recognizer created successfully\n");
  ak = loadEngineObject(AK_RES);
  printf(" . AlphabetKnowledge resource loaded successfully\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, recognizer, ak));
  printf(" . AlphabetKnowledge resource attached successfully to the recognizer\n");
  lk = loadEngineObject(LK_TEXT_RES);
  printf(" . Main LinguisticKnowledge resource loaded successfully\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, recognizer, lk));
  printf(" . Main LinguisticKnowledge resource attached successfully\n");
  if (strstr(LANG, "en") == NULL)
  {
    const char* path;
    if (strstr(LANG, "zh") != NULL || strstr(LANG, "ja") != NULL || strstr(LANG, "ko") != NULL)
      path = LK_ENGLISH_SECONDARY_2_RES;
    else if (strstr(LANG, "hi") != NULL)
      path = LK_ENGLISH_SECONDARY_3_RES;
    else
      path = LK_ENGLISH_SECONDARY_1_RES;

    lkEnglishSecondary = loadEngineObject(path);
    printf(" . Secondary English LinguisticKnowledge resource loaded successfully\n");
    verify(VO_IFACE(engine, IAttachTarget)->attach(engine, recognizer, lkEnglishSecondary));
    printf(" . Secondary English LinguisticKnowledge resource attached successfully\n");
  }
  printf("\n");

  printf("4) creating a new WriterAdaptationCollector\n");
  collector = voCreateObject(engine, VO_WriterAdaptationCollector);
  verify(collector != NULL);
  printf(" . collector created successfully\n");
  printf("\n");

  printf("5) attaching the profile to the collector\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, collector, profile));
  printf(" . profile attached successfully to the collector\n");
  printf("   . the collector will store writer data into the profile for further training\n");
  printf("\n");

  printf("6) attaching the collector to the recognizer\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, recognizer, collector));
  printf(" . collector attached successfully to the recognizer\n");
  printf("   . the recognizer will transfer incoming writer data to the collector\n");
  printf("\n");

  printf("7) attaching the profile to the recognizer\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, recognizer, profile));
  printf(" . profile attached successfully to the recognizer\n");
  printf("   . the recognizer will use existing writer data to adapt the recognition process\n");
  printf("\n");

  printf("8) sending digital ink to the recognizer\n");
  input = voCreateFloatStructuredInput(engine);
  verify(input != NULL);
  {
    struct AppContext context = { engine, input };
    FILE* file = fileOpen(DATA_DIR "/sentence.ink", "r");
    if (file != NULL)
    {
      verify(parseInk(file, parseInkCallback, &context) != EXIT_FAILURE);
      fclose(file);
    }
  }
  verify(VO_IFACE(engine, IInputConsumer)->setSource(engine, recognizer, input));
  printf(" . digital ink successfully sent to the recognizer\n");
  printf("\n");

  printf("9) running the conjoint handwriting recognition/adaptation process\n");
  initializeTextUIProgressBar(&progressBar);
  verify(VO_IFACE(engine, IRunnable)->run(engine, recognizer, updateTextUIProgressBar, &progressBar));
  finalizeTextUIProgressBar(&progressBar);
  printf(" . handwriting recognition/adaptation performed successfully\n");
  printf("\n");

  printf("10) committing the collected data\n");
  verify(VO_IFACE(engine, IWriterAdaptationCollector)->commit(engine, collector));
  printf(" . collected data committed successfully\n");
  printf("\n");

  printf("11) creating and configuring a new WriterAdaptationTrainer\n");
  trainer = voCreateObject(engine, VO_WriterAdaptationTrainer);
  verify(trainer != NULL);
  printf(" . trainer created successfully\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, trainer, ak));
  printf(" . AlphabetKnowledge resource attached successfully to the trainer\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, trainer, profile));
  printf(" . profile attached successfully to the trainer\n");
  printf(" . trainer created and configured successfully\n");
  printf("\n");

  printf("12) running the training process on the collected data\n");
  initializeTextUIProgressBar(&progressBar);
  verify(VO_IFACE(engine, IRunnable)->run(engine, trainer, updateTextUIProgressBar, &progressBar));
  finalizeTextUIProgressBar(&progressBar);
  printf("   . the trainer updated the collected writer data; next recognition will adjust itself to the writer writing style\n");
  printf("\n");

  printf("13) storing the profile to disk\n");
  storeEngineObject(profile, PROFILE_FILE);
  printf(" . profile stored successfully to: ");
  printPath(stdout, PROFILE_FILE);
  printf("\n");
  printf("\n");

  printf("14) cleaning-up: releasing all created objects and destroying the Engine\n");
  verify(voReleaseObject(engine, input));
  verify(voReleaseObject(engine, trainer));
  verify(voReleaseObject(engine, collector));
  if (lkEnglishSecondary != NULL)
    verify(voReleaseObject(engine, lkEnglishSecondary));
  verify(voReleaseObject(engine, lk));
  verify(voReleaseObject(engine, ak));
  verify(voReleaseObject(engine, recognizer));
  verify(voReleaseObject(engine, profile));
  verify(voDestroyEngine(engine));
  printf(" . clean-up done successfully\n");
  printf("\n");

  printf("end of example:\n");
  printf(" you just learnt how to perform writer adaptation\n");

  // -- finalization -----------------------------------------------------------
  unloadMyScriptEngine();

  return EXIT_SUCCESS;
}
